src/hwperfmon.c

```diff
@@ -0,0 +1,325 @@
+/* Copyright 2020 SiFive, Inc */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#include <metal/hwperfmon.h>
+#include <stdint.h>
+
+/* Macro to generate code within a switch case */
+#define METAL_HWPERFMON_HANDLE_SWITCH(m)                                       \

```

Comment:
Some funky macro.  Not such a big fan of them, but don't have an alternative at the moment.